File: core/src/main/java/io/github/shaksternano/borgar/core/media/template/Template.java
package io.github.shaksternano.borgar.core.media.template;

import io.github.shaksternano.borgar.core.media.AudioFrameOld;
import io.github.shaksternano.borgar.core.media.ImageFrameOld;
import io.github.shaksternano.borgar.core.media.graphics.Position;
import io.github.shaksternano.borgar.core.media.graphics.TextAlignment;
import io.github.shaksternano.borgar.core.media.graphics.drawable.Drawable;
import io.github.shaksternano.borgar.core.media.readerold.MediaReader;

import java.awt.*;
import java.io.IOException;
import java.util.Optional;
import java.util.function.Function;

public interface Template {

    MediaReader<ImageFrameOld> getImageReader() throws IOException;

    MediaReader<AudioFrameOld> getAudioReader() throws IOException;

    String getFormat();

    String getResultName();

    int getImageContentX();

    int getImageContentY();

    int getImageContentWidth();

    int getImageContentHeight();

    Position getImageContentPosition();

    int getTextContentX();

    int getTextContentY();

    int getTextContentWidth();

    int getTextContentHeight();

    Position getTextContentPosition();

    TextAlignment getTextContentAlignment();

    Font getFont();

    Color getTextColor();

    Optional<Function<String, Drawable>> getCustomTextDrawableFactory();

    double getContentRotation();

    Optional<Shape> getContentClip() throws IOException;

    boolean isBackground();

    Optional<Color> getFill();
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/media/template/Template.kt
package io.github.shaksternano.borgar.core.media.template

import io.github.shaksternano.borgar.core.graphics.ContentPosition
import io.github.shaksternano.borgar.core.graphics.TextAlignment
import io.github.shaksternano.borgar.core.graphics.drawable.Drawable
import io.github.shaksternano.borgar.core.io.DataSource
import io.github.shaksternano.borgar.core.media.reader.AudioReader
import io.github.shaksternano.borgar.core.media.reader.ImageReader
import java.awt.Color
import java.awt.Font
import java.awt.Shape

interface Template {

    val media: DataSource
    val format: String
    val resultName: String

    val imageContentX: Int
    val imageContentY: Int
    val imageContentWidth: Int
    val imageContentHeight: Int
    val imageContentPosition: ContentPosition

    val textContentX: Int
    val textContentY: Int
    val textContentWidth: Int
    val textContentHeight: Int
    val textContentPosition: ContentPosition
    val textContentAlignment: TextAlignment

    val font: Font
    val textColor: Color
    val customTextDrawableSupplier: ((String) -> Drawable)?
    val contentRotationRadians: Double
    val isBackground: Boolean
    val fill: Color?
    val forceTransparency: Boolean

    suspend fun getImageReader(): ImageReader

    suspend fun getAudioReader(): AudioReader

    suspend fun getContentClip(): Shape?
}
